packages/block-editor/src/components/provider/use-block-entity-sync.js

```diff
@@ -0,0 +1,163 @@
+/**
+ * External dependencies
+ */
+import { last, noop } from 'lodash';
+
+/**
+ * WordPress dependencies
+ */
+import isShallowEqual from '@wordpress/is-shallow-equal';
+import { useEffect, useRef } from '@wordpress/element';
+import { createHigherOrderComponent } from '@wordpress/compose';
+
+export function useBlockEntitySync( {
+	registry,
+	value: controlledBlocks,
+	clientId = null,
+	onChange = noop,
+	onInput = noop,
+	selectionStart: incomingSelectionStart,
+	selectionEnd: incomingSelectionEnd,
```

Comment:
What are  these selection things for?

Reply:
They've been here for a while. It integrates selection state with undo history.

Reply:
ok. I guess my question is why  it needs to be an API of the hook? can we retrieve these from the store?

Reply:
They come from the same place the blocks come. They need to be synced between `core/editor` and `core/block-editor`.

Reply:
ok got it, so props too. Why don't we need these in InnerBlocks too? I guess I need to dive in to understand how these are used in editor

Reply:
I think we just haven't set up selection state for controlled inner blocks yet (???) I mean, these props are passed directly to `<BlockEditorProvider>` wherever we set it up. If you notice, `selectionStart` are also retrieved in the state at some point below. But there is a difference between that and `this.props.selectionStart`. (At least, this is how the original code was set up)

I debated putting these values in this hook. Maybe there should be a separate effect in `<BlockEditorProvider />` which updates these values? I copied them over because the logic was intertwined with some of the controlled block sync stuff

Reply:
>I think we just haven't set up selection state for controlled inner blocks yet

Yeah, it would need to be passed through `useEntityBlockEditor`.

Reply:
I think we can fix this in a follow-up 👍 I think it just messes with selection state when undoing changes in a template part